### PR TITLE
Remove GCC and zlib from the container

### PR DESCRIPTION
**Pull request by bgruening**
This one will remove a lot of dependencies, gcc, g++, zlib. This is to force new recipes to use stricter build- and run-time dependencies.

This PR is mainly to start a discussion. I once tried to remove zlib and fix all old recipes and failed because it can be really hard for some recipes or the URL from these old packages were not available.

### CODE REVIEW

**Comment by johanneskoester** [comment]
I think this is a good idea. I don't see any downsides since gcc llvm and zlib are available as conda packages. The rebuilding issue is kind of independent of this, because rebuilds would also fail if the deps were in the docker container. Hence, I'm in favor of merging this as soon as possible. What do you think @bioconda/core?

**Comment by bgruening** [comment]
This would be a nice first step in the direction of unifying the build systems with other community-efforts, like Omnia or conda-forge. Ideally we depend on https://github.com/phusion/holy-build-box on some point or similar systems.

**Comment by daler** [comment]
I think updating the docker container is a good idea, especially to get new recipes to be more explicit with dependencies.

I expect transitioning all existing packages over will be straightforward if time-consuming -- essentially bump all build versions and see what breaks.

holy-build-box seems like a good direction for the future, but it's not clear to me how to map the exisiting way conda builds things to the methods outlined in the holy-build-box docs.

**Comment by bgruening** [comment]
@johanneskoester @daler do we want to merge this and rebuild the image?

**Comment by daler** [comment]
@bgruening let me do some local testing -- I want to get a feel for if/how much things may break. Then I'll make sure notes are in place in `GUIDELINES.md` for when the container goes into production.

**Comment by daler** [comment]
@bgruening, gcc is still making it into the container. I suspect this is because `devtoolset-2-gcc-gfortran` is still being installed, and the two `-devel` libs (`db4-devel` and `mesa-libGLU-devel`) are likely pulling it in as a dependency.

Looks like we previously agreed to keep X11 and the mesa-libGLU-devel (https://github.com/bioconda/bioconda-recipes/issues/624#issuecomment-172196183).

Remaining questions:

1. Do you think we still need `devtoolset-2-gcc-gfortran`?
2. Do we still need `db4-devel`? Comment in Dockerfile says this is needed for `perl-db-file` recipe, but if I remember correctly this was from the days when we were taking the easy way out (add stuff to Dockerfile rather than make a recipe).
3. What do you think about `mv /usr/bin/gcc /opt/gcc` in the Dockerfile? That way even if gcc makes it in as an inadvertent dependency, it won't be found by travis-ci

**Comment by daler** [comment]
OK @bgruening, getting closer . . . but I'm getting zlib errors.

After removing gcc-gfortran and moving `/usr/bin/gcc` (as of b7ad75c), I built the container:

```
(cd scripts && docker build -t no-gcc)
```

then fixed the `samtools` recipe:

```
diff --git a/recipes/samtools/meta.yaml b/recipes/samtools/meta.yaml
index 9771c29..4b17afa 100644
--- a/recipes/samtools/meta.yaml
+++ b/recipes/samtools/meta.yaml
@@ -3,7 +3,7 @@ package:
   version: "1.3"
 
 build:
-  number: 1
+  number: 2
   skip: False
 
 source:
@@ -12,6 +12,7 @@ source:
 
 requirements:
   build:
+  - gcc
   - ncurses
   - zlib
   run:
```

then tried building samtools:

```
docker run --net="host" -it -v `pwd`:/bioconda-recipes no-gcc -v  --packages samtools
```

however I'm getting a zlib error, even though apparently I have `zlib` as a build dependency:

```
The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    gmp-5.1.2                  |                2         594 KB
    ncurses-5.9                |                4         676 KB
    isl-0.12.2                 |                0         1.1 MB
    mpfr-3.1.2                 |                0         407 KB
    cloog-0.18.0               |                0         617 KB
    mpc-1.0.1                  |                0          61 KB
    gcc-4.8.5                  |                3        65.8 MB
    ------------------------------------------------------------
                                           Total:        69.2 MB

The following NEW packages will be INSTALLED:

    cloog:   0.18.0-0
    gcc:     4.8.5-3 
    gmp:     5.1.2-2 
    isl:     0.12.2-0
    mpc:     1.0.1-0 
    mpfr:    3.1.2-0 
    ncurses: 5.9-4   
    zlib:    1.2.8-0 

gcc install OS matches gcc build OS: Skipping post-link portability fixes.
Removing old work directory
Source cache directory is: /anaconda/conda-bld/src_cache
Downloading source to cache: samtools-1.3.tar.bz2
Downloading https://github.com/samtools/samtools/releases/download/1.3/samtools-1.3.tar.bz2
Success
Extracting download
Package: samtools-1.3-2
source tree in: /anaconda/conda-bld/work/samtools-1.3
+ make install prefix=/anaconda/envs/_build
Makefile:99: config.mk: No such file or directory
gcc -g -Wall -O2 -D_FILE_OFFSET_BITS=64 -D_LARGEFILE64_SOURCE -I. -Ihtslib-1.3  -c -o bam_index.o bam_index.c
gcc -g -Wall -O2 -D_FILE_OFFSET_BITS=64 -D_LARGEFILE64_SOURCE -I. -Ihtslib-1.3  -c -o bam_plcmd.o bam_plcmd.c
gcc -g -Wall -O2 -D_FILE_OFFSET_BITS=64 -D_LARGEFILE64_SOURCE -I. -Ihtslib-1.3  -c -o sam_view.o sam_view.c
gcc -g -Wall -O2 -D_FILE_OFFSET_BITS=64 -D_LARGEFILE64_SOURCE -I. -Ihtslib-1.3  -c -o bam_cat.o bam_cat.c
In file included from bam_cat.c:42:0:
htslib-1.3/htslib/bgzf.h:34:18: fatal error: zlib.h: No such file or directory
 #include <zlib.h>
                  ^
compilation terminated.
make: *** [bam_cat.o] Error 1
Command failed: /bin/bash -x -e /bioconda-recipes/recipes/samtools/build.sh
```

This had worked before presumably because zlib-devel was in the docker container. Any ideas on what's happening here?

**Comment by kyleabeauchamp** [comment]
One (low priority) thing I just thought of is CUDA support, where (AFAIK, see e.g. http://developer.download.nvidia.com/compute/cuda/7.5/Prod/docs/sidebar/CUDA_Installation_Guide_Linux.pdf page 6) CUDA will require having a system-wide GCC install.  I wonder if there are other situations like that.

**Comment by daler** [comment]
I have no experience with CUDA, but based on that doc it looks like it only needs `gcc` to be on the path. I think it should work if `gcc` is included in the build requirements of the recipe.

**Comment by kyleabeauchamp** [comment]
Possibly.  I'm not 100% sure.  The common way to install CUDA is via either apt or yum, so presumably that would force-install the requirements.  There is also a shell script installer, but I'm not sure how well it behaves with non-root PATHs in play.  

TLDR it would be an experiment.

**Comment by bgruening** [comment]
@daler For some unknown reason zlib is can often not be found. Have you tried to add 

```
export CFLAGS="-I$PREFIX/include"
export LDFLAGS="-L$PREFIX/lib"
```

> 1. Do you think we still need `devtoolset-2-gcc-gfortran`?

No.

> 2. Do we still need `db4-devel`? Comment in Dockerfile says this is needed for `perl-db-file` recipe, but if I remember correctly this was from the days when we were taking the easy way out (add stuff to Dockerfile rather than make a recipe).

Don't think so, we should work on a package level to fix those things. Otherwise it will affect other packages without knowing.

> 3. What do you think about `mv /usr/bin/gcc /opt/gcc` in the Dockerfile? That way even if gcc makes it in as an inadvertent dependency, it won't be found by travis-ci

As last resort this is ok, but I would rather figure our which package add the compiler.

Thanks for helping here!

**Comment by daler** [comment]
Thanks, but trying 
```
export CFLAGS="-I$PREFIX/include"
export LDFLAGS="-L$PREFIX/lib"
```
 in samtools/build.sh didn't work -- same error as above. Any other ideas?

As far as packages, I narrowed it down to just `autotools-latest` that is pulling in `gcc` as a dependency.

**Comment by bgruening** [comment]
One last idea; Can you import `- pkg-config` as build dep? I will have time tm to try by myself, but zlib can be really ugly. I wish I would know why this is so special.

**Comment by bgruening** [comment]
@daler;

The following is working for me:
```
sed -i.bak 's/^CPPFLAGS =$//g' Makefile
sed -i.bak 's/^LDFLAGS  =$//g' Makefile

sed -i.bak 's/^CPPFLAGS =$//g' htslib-1.3/Makefile
sed -i.bak 's/^LDFLAGS  =$//g' htslib-1.3/Makefile

export CPPFLAGS="-I$PREFIX/include"
export LDFLAGS="-L$PREFIX/lib"
```

I will patch samtools to make this unnecessary with a new upstream samtools release. Do you want me to patch bioconda samtools before or after merging this one?

**Comment by daler** [comment]
Ah, nice. Thanks.

As far as patching, whatever's convenient for you.

**Comment by bgruening** [comment]
@daler: https://github.com/bioconda/bioconda-recipes/pull/1294
Do we want to patch all other versions as well?

**Comment by daler** [comment]
@bgruening now that I think about it, might as well -- otherwise we'd need documentation on which samtools versions used which zlib

**Comment by bgruening** [comment]
This takes longer, no proper internet here. Feel free to merge this one. I will fix samtools later.

**Comment by acaprez** [comment]
On the subject of CUDA, I think that's basically a non-starter anyway as long as the build environments are RHEL5-based.  CUDA is pretty tightly coupled to the particular system default compiler and glibc version for each platform.  RHEL5 hasn't been supported since toolkit version 6.0.

**Comment by kyleabeauchamp** [comment]
Well, we've been discussing upgrading to the holy-build-box, which is based on a more recent version of RHEL.

**Comment by kyleabeauchamp** [comment]
I take that back, it's based on CENTOS 5.  However, my colleagues have been able to get CUDA working on a variant of the holy build box: https://github.com/omnia-md/omnia-build-box/blob/master/Dockerfile

**Comment by acaprez** [comment]
Interesting.  Still, it's not an officially supported configuration by Nvidia, which may end up causing complications.  And then there's the issue of how to do the CI testing.  AFAIK, neither travis nor anaconda provide GPU-equipped build hosts.

**Comment by bgruening** [comment]
@daler ready to merge?

**Comment by daler** [comment]
Sorry, swamped with other work -- haven't had time to test anything beyond samtools. Let me spot-check a couple and then I will merge.

**Comment by daler** [comment]
@bgruening  `r-spp` works fine after adding `gcc` (doesn't depend on zlib I guess).

However I can't get `bedtools` or `trinity` to work, they both fail to find zlib. I'm trying to export CPPFLAGS and LDFLAGS as you suggest, and looking for where they are defined in their respective sources doesn't turn up anything obvious that needs to be `sed`ed out.

My failure to fix these packages is likely because I don't know enough about C and C++ libraries. Any other suggestions for what to try? 

I suspect these problems will be solved by patching the source rather than adding zlib back into the container. But before merging this, I think it would be nice to have a list in GUIDELINES.md of things to try, or at least some troubleshooting tips, to help retrofit existing recipes to build on this new container. For example:

- make sure LDFLAGS and CPPFLAGS are not defined in any Makefiles
- try adding `export CPPFLAGS="-I$PREFIX/include"` and   `export LDFLAGS="-L$PREFIX/lib"` to `build.sh`
- ...?

**Comment by kyleabeauchamp** [comment]
FWIW, is anyone else concerned about having to patch the source to get basic zlib dependencies to work?  I realize there might not be a good solution (e.g. real bugs in upstream packages) but still might merit some discussion.

**Comment by johanneskoester** [comment]
@yesimon, do you have any comments on this? I know you are pretty experienced with C/C++, right?

**Comment by acaprez** [comment]
@daler For `bedtools`, adding `export CPATH=${PREFIX}/include` in the build script seemed to work when I tried it.

**Comment by bgruening** [comment]
@daler https://github.com/bioconda/bioconda-recipes/issues/1336 fixes bedtools and trinity for me, but only with the new proposed Docker file. It fails with the old Dockerfile.

@kyleabeauchamp I tend to think that all packages that do not strictly depend on zlib or gcc are broken. For example I can not installed them in mulled or on our cluster without zlib.
For me it would be fine if we merge this PR and from now on all packages needs to declare zlib and gcc. Old packages can be fixed but we don't need to.

**Comment by johanneskoester** [comment]
@bgruening I totally agree!

**Comment by yesimon** [comment]
On phone..

I think that removing gcc would cause problems but requiring zlib is okay. Removing gcc would require everyone to put gcc in deps, but conda is not great at handling different versions of libraries. Since gcc in conda is being updated, we'll eventually have a mishmash of packages built with different gcc versions which can cause hard to diagnose problems when mixing libraries. To fix this we can manually do full rebuilds with new gcc ABI changes but this requires strong coordination with end users like in arch Linux. I think that's unreasonable for the conda audience. It would be much easier having a set gcc version for all packages in container, with exceptions for a small subset of new packages requiring new features.

For zlib, the problem is most upstream authors just assume zlib will be available and have the right version. Zlib is very stabilized and I think last released was 3 years ago so version is no issue. The only issue is fixing make files and such, which would be either upstream (ideal), or patches. Both require a bit of extra work, but since many scientific cluster environments seem to be missing basic packages like zlib this is a good targeted intervention.

**Comment by bgruening** [comment]
>I think that removing gcc would cause problems but requiring zlib is okay. Removing gcc would require everyone to put gcc in deps, but conda is not great at handling different versions of libraries. Since gcc in conda is being updated, we'll eventually have a mishmash of packages built with different gcc versions which can cause hard to diagnose problems when mixing libraries. To fix this we can manually do full rebuilds with new gcc ABI changes but this requires strong coordination with end users like in arch Linux. I think that's unreasonable for the conda audience. It would be much easier having a set gcc version for all packages in container, with exceptions for a small subset of new packages requiring new features.

We have the same problem if we do not specify it, just on a different level - inside the container, isn't it? If a program needs a never version (or even older) of GCC than provided in the container we have the same problem. Maybe even worse because we only offer one version of GCC. I guess the best thing we can do today is to hope that Continuum comes up with a nice solution for this problem and annotate our dependencies as explicit as possible.

**Comment by yesimon** [comment]
Yes but the version is changed less frequently than conda gcc, and older is typically better for compatibility. Two other things: conda-recipes is removing gcc from their builder and using implicit, and in fact we should probably match their version. Gcc is mainly used for build and we don't gain as much for end users when specifying versus say a library like zlib.

**Comment by bgruening** [comment]
> Yes but the version is changed less frequently than conda gcc, and older is typically better for compatibility. 

Currently, there is only one GCC (4.8.x) in the default channel, with 1 point release.
Older is better, if your program compiles with older GCC versions, if you rely on new C++ features your are lost ...

> Two other things: conda-recipes isnt removing GCC from their builder and using implicit, 

Interesting do you have any reference here?


> and in fact we should probably match their version. 

I guess this is another reason we should share the same build infrastructure. But I don't see where this helps in any of the mentioned problems above. As soon as they change the GCC, we are bound to do as well.

> GCC is mainly used for build and we don't gain as much for end users when specifying versus say a library like zlib.

What do we loose if we specify it? Would it be ok to bind a recipe to a strict GCC version (<=4.8.5)?
What is with libgcc, which is needed as runtime dependency quite a lot.

My main argument is that it is more transparent for everyone if we specify gcc/llvm/libgcc instead of hide this detail in the build-box. But I agree `zlib` is way more important than GCC in this PR.

**Comment by johanneskoester** [comment]
I also favor to have both as dependencies. Also, we can pin recipes to the major version of gcc (see my recent PR on pinning for boost). Is it true that you cannot depend on a library that was compiled with a different gcc? I thought gcc is pretty much in a mode where only internals are changed, and it should not affect the ABI of a library, shouldn't it (just guessing here)?

**Comment by yesimon** [comment]
No we cannot rely on gcc being fully compatible across major versions for built libraries even though things typically work (when things don't work it's bad news). Pinning gcc version to what we're currently using (4.8 I think) is desirable, although I don't view it much differently from storing this in environment variable vs Dockerfile.

**Comment by johanneskoester** [comment]
The difference is that the resulting package can depend on a pinned libgcc 4.8 then. Hence, two packages build against different gcc versions cannot be combined. Further, depending on gcc allows us to define constraints regarding the version. E.g., there can be packages that need a minimum gcc version.

**Comment by bgruening** [comment]
Totally agree with @johanneskoester and it is explicit rather than implicit, which is good :)

**Comment by bgruening** [comment]
@yesimon do you feel strong about this detail, aka is this a `-1`?

**Comment by yesimon** [comment]
@bgruening I think it would be fine to pin as long as we continuously new recipes for possible `gcc`/`libgcc` discrepancies!

**Comment by bgruening** [comment]
Ok, so let's pin it. @johanneskoester do you want to work on that or do you me to change this PR with the pinning stuff included?

**Comment by johanneskoester** [comment]
Mhm, maybe we should have a look at this before we decide: https://github.com/conda-forge/staged-recipes/issues/436

I also asked the other channels for their general opinion here (so far no answer): https://github.com/conda-forge/staged-recipes/issues/299#issuecomment-212797647

**Comment by johanneskoester** [comment]
It seems like continuum failed to pin some gcc dependencies. As updates have arrived in the main channel recently, at least g++ is broken in the default channel now. This has been fixed in conda-forge by rebuilding some of the dependencies. This is a good reason to add conda-forge as a channel in our build environment and to update the docs to include conda-forge as a channel we depend on. Of course, there is the risk of breaking something, but after the discussions [here](https://github.com/conda-forge/staged-recipes/issues/299], I am quite confident that there should be no negative impact for us. What do you think, @bioconda/core?

**Comment by daler** [comment]
I agree that adding conda-forge as a channel dependency is not a bad thing. It may increase complexity somewhat -- for example, we should incorporate a check in the build scripts to ensure a new bioconda recipe doesn't unintentionally shadow a conda-forge recipe -- but I think the benefits of agility outweigh that.

(sorry, I haven't been following the various gcc discussions closely enough to make any informed comments on that topic)

**Comment by bgruening** [comment]
:+1: from my side!

**Comment by yesimon** [comment]
The only question I have is ownership of packages between `bioconda` and `conda-forge`. Especially since both projects are expanding quickly, there might need to be coordination between the teams to send packages to the right repository.

**Comment by johanneskoester** [comment]
Indeed. We need to set up some rules at least on our side. However, given that there seem to be some problems in the conda-forge channel (the gcc issue), let's just wait a bit until things have stabilized on their side. An alternative is still to rely on conda-recipes instead. I know, it is slower, but since stuff goes into the default channel, it is presumably more stable.

**Comment by bgruening** [comment]
How do we want to preceded here. Can we remove zlib and gcc from the build environment?
Or should I split this PR into two separate ones?

**Comment by jakirkham** [comment]
> I also asked the other channels for their general opinion here (so far no answer): conda-forge/staged-recipes#299 (comment)

I have tried to answer your question in this [comment]( https://github.com/conda-forge/staged-recipes/issues/299#issuecomment-216072561 ), @johanneskoester. If you have more questions, feel free to ask. Sorry for the delay.

**Comment by bgruening** [comment]
I guess this one can be closed.

**Comment by bgruening** [comment]
I guess this one can be closed.
